crates/ide-db/src/path_transform.rs

```diff
@@ -327,8 +327,36 @@ impl Ctx<'_> {
                     ted::replace(path.syntax(), subst.clone_subtree().clone_for_update());
                 }
             }
+            hir::PathResolution::SelfType(imp) => {
+                let ty = imp.self_ty(self.source_scope.db);
+                let ty_str = &ty
+                    .display_source_code(
+                        self.source_scope.db,
+                        self.source_scope.module().into(),
+                        true,
+                    )
+                    .ok()?;
+                let ast_ty = make::ty(&ty_str).clone_for_update();
+
+                if let Some(adt) = ty.as_adt() {
+                    if let ast::Type::PathType(path_ty) = &ast_ty {
+                        let found_path = self.target_module.find_use_path(
+                            self.source_scope.db.upcast(),
+                            ModuleDef::from(adt),
+                            false,
+                        )?;
+
+                        if let Some(qual) = mod_path_to_ast(&found_path).qualifier() {
+                            let res = make::path_concat(qual, path_ty.path()?).clone_for_update();
+                            ted::replace(path.syntax(), res.syntax());
+                            return Some(());
+                        }
+                    }
+                }
+
+                ted::replace(path.syntax(), ast_ty.syntax());
```

Comment:
We probably wanna prefer keeping `Self` if it does not need to be replaced. So a check if the source and target scopes have the same `Self` (and in that case, skip this transformation here entirely). We can give `Ctx` a new bool field for that and set it accordingly in `build_ctx` by giving `SemanticsScope` a function like `fn has_same_self_type(&self, other: &SemanticsScope<'_>) -> bool` which just checks if the top `ImplDefScope`s of each scope's resolver has the same def in them.